#167942054 User with expired token should be redirected to login page

Manorlds-Eaglespark opened the pull request:
## Description ##
- A user should be redirected back to the login page when the JWT token they got from the server is expired. This will resolve #51 
- The navbar does not scroll out of view 

## Type of change ##
Please select the relevant option
- [ ] Bug fix(a non-breaking change which fixes an issue)
- [x] New feature(a non-breaking change which adds functionality)
- [ ] Breaking change(fix of a feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested? ##
Please describe the tests that you ran to verify your changes
- [ ] End to end test
- [ ] Integration test
- [x] unit test

## Checklist: ##
- [x] My code follows the style guidelines of this project
- [x] I have linted my code prior to submission
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes

## How to test this PR
When after a set time, you are redirected to the login page to sign in again due to an expired fixed time. In your local deployment, you can use this expired token through your code inspection tool in your browser. 
 ```eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VybmFtZSI6IkFoZWJ3YTEiLCJlbWFpbCI6ImNyeWNldHJ1bHlAZ21haWwuY29tIiwiZXhwIjoxNTUxNzc2Mzk0fQ.PFimaBvSaxR_cKwLmeRMod7LHkhNTcem22IXTrrg7Ko```


## Related Pivotal Tracker stories ##
[#167942054](https://www.pivotaltracker.com/story/show/167942054)

Manorlds-Eaglespark committed:
feat(redirect-to-login-page)User with expired token should be redirected to login page
- user with expired token is redirected automatically
[Delivers #167942054]

Manorlds-Eaglespark committed:
feat(redirect-to-login-page)User with expired token should be redirected to login page
- user with expired token is redirected automatically
[Delivers #167942054]

Tevinthuku commented:
@Manorlds-Eaglespark  you have a failing test on `circle ci`
 
![image](https://user-images.githubusercontent.com/12128153/63519136-13d3e500-c4fb-11e9-9631-0b33f3b3d3b9.png)

Tevinthuku reviewed (commented):
Ive pasted in your token from the PR description, Im getting the following error

![image](https://user-images.githubusercontent.com/12128153/63519552-dd4a9a00-c4fb-11e9-9149-e4fa39b1a238.png)
and Im not being redirected to the login page

neelxie reviewed (commented):
<img width="1202" alt="image" src="https://user-images.githubusercontent.com/32802973/63522737-995a9380-c501-11e9-8295-e15200cdb325.png">

Everything seems to  be good apart from your build, as it throws me out when I try to access my profile plus if you could kindly please outline the protected routes any one could access with the token you have provided in the PR to make this review just a few seconds long, good work bro.

Tevinthuku commented on src/app/helpers/jwt.interceptor.spec.ts:
```diff
@@ -1,45 +1,76 @@
   });
 
+  const setUp = (authService: AuthService) => {
+    spy = jasmine.createSpyObj('HttpHandler', ['handle']);
```
For uniformity, we can have this jasmine spy in the `src/app/helpers/spies.ts` file

Manorlds-Eaglespark replied:
I like this, this also was resolved. Thanks @Tevinthuku

Tevinthuku commented on src/app/helpers/jwt.interceptor.ts:
```diff
@@ -1,16 +1,22 @@
+import { Injectable } from '@angular/core';
+import { HttpEvent, HttpHandler, HttpInterceptor, HttpRequest } from '@angular/common/http';
+import { Observable } from 'rxjs';
+import { AuthService } from '../../app/services/auth.service';
```
@Manorlds-Eaglespark , we can use `absolute path` here

Manorlds-Eaglespark replied:
This has been resolved, kindly take another look

Tevinthuku replied:
not yet @Manorlds-Eaglespark

Manorlds-Eaglespark replied:
Please check, thanks @Tevinthuku

Tevinthuku commented:
@Manorlds-Eaglespark , 🤔 , could I be missing something since I'm not seeing where exactly the redirect to the login page is happening

wangonya commented on src/app/helpers/jwt.interceptor.spec.ts:
```diff
@@ -1,45 +1,76 @@
+import { HttpHandler, HttpRequest } from '@angular/common/http';
+import { of } from 'rxjs';
+import { JwtInterceptor } from './jwt.interceptor';
+import { AuthService } from '../services/auth.service';
```
Let's make use of absolute imports.

Manorlds-Eaglespark replied:
Thanks, this has been corrected.

Tevinthuku replied:
Hey @Manorlds-Eaglespark , this hasn't yet been fixed

Manorlds-Eaglespark replied:
Please check again, thanks  @Tevinthuku

alexxsanya reviewed (approved):
@Manorlds-Eaglespark. I have checked it out and is working as state.

Tevinthuku commented:
Hey @Manorlds-Eaglespark , nice work. kindly fix the import issues so that this can be approved

Tevinthuku commented:
> <img alt="image" width="1202" src="https://user-images.githubusercontent.com/32802973/63522737-995a9380-c501-11e9-8295-e15200cdb325.png">
> 
> Everything seems to be good apart from your build, as it throws me out when I try to access my profile plus if you could kindly please outline the protected routes any one could access with the token you have provided in the PR to make this review just a few seconds long, good work bro.

@neelxie ,This is what it's supposed to do. It seems your token expired, and if it does so in a protected route, you get redirected to the login page, once you log in successfully, you get redirected back to the page you were in

check the url of the page, its `/login?next=profile`, which means you will be redirected to the `profile` page when you successfully login

Tevinthuku commented on src/app/helpers/jwt.interceptor.spec.ts:
```diff
@@ -1,45 +1,69 @@
-import {JwtInterceptor} from './jwt.interceptor';
+import { TestBed } from '@angular/core/testing';
+import { HttpClientTestingModule } from '@angular/common/http/testing';
+import { httpHandlerSpy, httpRequestSpy } from './spies';
```
@Manorlds-Eaglespark , kindly fix this to use `absolute imports`

Manorlds-Eaglespark replied:
Thanks, @Tevinthuku, the relative import was removed.

Niyitangasam commented on src/app/helpers/jwt.interceptor.spec.ts:
```diff
@@ -1,45 +1,69 @@
+    service.intercept(httpRequestSpy, httpHandlerSpy)
       .subscribe(
         result => {
           console.log('good', result);
```
@Manorlds-Eaglespark , please remove this console.

Niyitangasam commented on src/app/helpers/jwt.interceptor.spec.ts:
```diff
@@ -1,45 +1,69 @@
+
+  authServiceData = {
+    currentUser: {
+      token: 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6MTAsImVtYWlsIjoib2dyZUB2aW1haWwyNC5jb20iLCJleHAiOjE1NjY2MzQxMDB9.5u2ylra-VwFb5qDjVsTVfcXPlGZlr9bGg54YyCwvkDA'
```
@Manorlds-Eaglespark , I am not understanding why is this hardcoded token here. Can you please help to understand?

patrickf949 reviewed (approved):
LGTM

wangonya reviewed (commented):
The feature works as described 👍🏾
We had talked about having a message that shows the user that their token/session had expired after they were redirected so they're not left wondering what happened. Did you speak to @kdaglas about it?

bl4ck4ndbr0wn reviewed (approved):
Found some issues after.

bl4ck4ndbr0wn reviewed (dismissed):
There is no action set for when a token is expired or invalid.

Manorlds-Eaglespark commented:
> There is no action set for when a token is expired or invalid.

Am using an expired token in my setup, which is the essence of the feature. @bl4ck4ndbr0wn

Tevinthuku reviewed (approved):
Nice job working on this @Manorlds-Eaglespark

Niyitangasam reviewed (approved):
Good job, tested it and work as described.

Victoradukwu reviewed (changes requested):
Nice job, @Manorlds-Eaglespark . A few things to note:
> It would be nice if the user is automatically taken to the page they were attempting to access before login
> This task also involves fixing the navbar that scrolls out of view and changing the hardcoded year in the footer--see the story

Victoradukwu reviewed (approved):
Works as expected

JamesMudidi reviewed (changes requested):
Hello @Manorlds-Eaglespark. great work but I would suggest you run `npm run lint` so that you can fix the linting issues on the changes you made

JamesMudidi commented on src/app/components/navbar/navbar.component.spec.ts:
```diff
@@ -1,43 +1,37 @@
 import { resetSpies } from 'src/app/helpers/social.spies';
 import { mockProfileResponse } from 'src/app/shared/mocks';
 import { of } from 'rxjs';
+import { By } from '@angular/platform-browser';
```
I see we have imported `By` here but I don't see where we are using it

JamesMudidi commented on src/app/components/navbar/navbar.component.spec.ts:
```diff
@@ -53,4 +47,19 @@ describe('NavbarComponent', () => {
+}));
+
+
+it('should create the aeepp --', async(() => {
```
Is this meant to be aeepp or app

Manorlds-Eaglespark replied:
This was a reverse test for an if statement. I have changed the naming. Thanks.

JamesMudidi commented on src/app/helpers/jwt.interceptor.spec.ts:
```diff
@@ -24,24 +34,37 @@ describe('JWTInterceptor', () => {
-    jwtInterceptor.intercept(httpRequestSpy, httpHandlerSpy)
+    service.intercept(httpRequestSpy, httpHandlerSpy)
       .subscribe(
         result => {
```
this result is declared but the value is never read

Manorlds-Eaglespark replied:
Some changes have come up due to rebasing.

JamesMudidi reviewed (dismissed):
you can also run the `format document` function in VSCode to help you tidy up the code for easier reading.

Manorlds-Eaglespark commented:
> you can also run the `format document` function in VSCode to help you tidy up the code for easier reading.

This has been done.

Manorlds-Eaglespark commented:
> There is no action set for when a token is expired or invalid.

There is, kindly look closely the feature is about redirection only when a token is expired.

kdaglas reviewed (approved):
LGTM

patrickf949 committed:
feat(redirect-to-deposits): redirect users to deposits page upon successful payment (#53)

- add route of deposits to the pin-validate service
- make sure pins are private
- add validation for input fields

[delivers #168242071]

dorothyas committed:
#165880984 Retrieve Client Reviews  (#47)

* ft-client-reviews: retriving-client-reviews
-
[Finishes #165880984]

* ft(client-reviews):get-client-reviews

- create client reviews component
- route to client details page
- improve UI design

[Finishes #165880984]

*  ft(client reviews): get reviews on client

- add tests

[Finishes #165880984]

shemaeric committed:
bug(Fix tests): Fix Financial Information tests (#62)

- Fix broken financial tests after codebase refactor
[Delivers #168408105]

Niyitangasam committed:
#168049088 Enable a buyer to review a client (#42)

* ft(review): Add a review to a client

- Add functionality to enables a buyer to add reviews to a client
- Add tests

[Delivers #168049088]

* Refactor code to use common spie in test

* Fix validations error

* Add a redirect url after adding a review

* Add test for routes

* Fix failing tests after rebasing

* Increase coverage

Manorlds-Eaglespark committed:
merge branch 'ft-user-expired-token-redirected-to-login-page' of https://github.com/landvilleng/landville-frontend into ft-user-expired-token-redirected-to-login-page

feat(redirect-to-login-page)User with expired token should be redirected to login page

- user with expired token is redirected automatically
[Delivers #167942054]
